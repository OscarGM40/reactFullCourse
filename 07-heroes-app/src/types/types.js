//los types van a ser los tipos para las acciones
export const types = {
   
   login: '[auth] login',
   logout:'[auth] logout'
   
}