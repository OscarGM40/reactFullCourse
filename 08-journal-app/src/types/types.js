//es mejor no meter las actions.types en duro

export const types = {

   login:'[Auth] Login',  
   logout:'[Auth] Logout',
   
   uiSetError:'[UI] Set Error',
   uiRemoveError:'[UI] Remove Error',

   uiStartLoading: '[UI] Start loading',
   uiFinishLoading: '[UI] Finish loading'


}