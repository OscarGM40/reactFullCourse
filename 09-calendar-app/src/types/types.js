
export const types = {
  uiOpenModal: "[ui] Open modal",
  uiCloseModal: "[ui] Close modal",

  eventSetActive: "[event] Set active",
  eventAddNew: "[event] Add new",
  eventClearActiveEvent: "[event] Clear active event",
  eventUpdate:"[event] Update event",
  eventDelete: "[event] Delete event",

  authCheckingStart: "[auth] Checking",
  authCheckingFinish: "[auth] Checking finish",
  authLoginSuccess: "[auth] Login success",
  authStartRegister: "[auth] Register start",
  authStartTokenRenewal: "[auth] Start token renewal",
  authLogout: "[auth] Logout",
  
}; 