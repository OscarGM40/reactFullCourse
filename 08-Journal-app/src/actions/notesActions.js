import Swal from "sweetalert2";
import { fileUpload } from "../helpers/fileUpload";
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";
import { db } from "../firebase/firebase-config";

// podriamos pasar el uid de cada usuario que tiene en Firebase como argumento pero también podemos usar una callback que nos proporciona thunk y nos devuelve el state.El nombre el que quiera,es el segundo argumento del return(getState aqui)
export const startNewNote = () => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid;
    
    const newNote = {
      title: "",
      body: "",
      date: new Date().getTime(), //mls actuales
    };
    
    try {
      /* Desde firebase 9 el método es addDoc y no add */
      const docRef = await db.collection(`${uid}/journal/notes`).add(newNote);
      // console.log(docRef);
      dispatch(activateNote(docRef.id,newNote))
      dispatch(addNewNote(docRef.id,newNote))
      
    } catch (error) {
      console.log(error);
    }
  };
};

export const addNewNote = (id,note) => ({
  type: types.notesAddNew,
  payload: {
    id,
    ...note,
  },
});

export const startSaveNote = ( note ) => {
  return async(dispatch,getState) => {
    // fijate que necesito el uid del usuario luego uso la cb
    const uid = getState().auth.uid;

    //  si no viene el url lo borro pues da error el undefined
    /* if(!note.url){
      delete note.url;
    } */
    
    // en realidad no necesito el delete,puedo cambiar el undefined por un null ya que puedo guardar un null  
    // saco las propiedades
     const { id, url=null,...rest} = note;
    //  console.log(url)
     /*  console.log(note,'note')
      console.log(id,'id') */

    /* const noteToFirestore = { ...note };
    delete noteToFirestore.id; */

    try {
      // db.collection es para trabajar en una coleccion
      // await db.collection(`path...`)
      // lo que quiero ahora es trabajar en un documento,
      await db.doc(`${uid}/journal/notes/${note.id}`)
      .update({ url, ...rest })  
      dispatch(refreshNote(note))
      Swal.fire("Saved",rest.title, 'success');
    } catch (error) {
      console.log(error)
    }

  }  
}

export const refreshNote = (note) => ({
  type:types.notesUpdate,
  payload: note
})

export const activateNote = (id,note) => ({
  type: types.notesActive,
  payload: { id, ...note }
}) // <- fijate que de esta forma las acciones sincronas devuelven un objeto luego return => ({})


export const startLoadingNotes = (uid) => {
  return async (dispatch) => {
    const notes = await loadNotes(uid)
    dispatch(setNotes(notes))
  }
}

// no haría falta romper la referencia porque estoy creando un nuevo arreglo y no apunto al de el state!
export const setNotes = (notes) => ({
  type:types.notesLoad,
  payload: notes
})

// como es una tarea asincrona es obvio qué debo usar
export const startUploading = (file) => {
  return async ( dispatch, getState ) => {
    const { active: activeNote } = getState().notes;

    Swal.fire({ 
      title: 'Uploading...',
      text: 'Please wait...',
      allowOutsideClick: false,
      onBeforeOpen: () => {
        Swal.showLoading()
      }
    });
    // llamo a mi helper que me devuelve el campo secure_url
    const fileUrl = await fileUpload(file);
    activeNote.url = fileUrl;
    // console.log(fileUrl,'fileUrl')

    dispatch(startSaveNote(activeNote))
    Swal.close();
    
  }
}

export const startDeleteNote = (id) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid;
    
    try {
      await db.doc(`${uid}/journal/notes/${id}`).delete();
      dispatch(deleteNote(id));
    } catch (error) {
      console.log(error);
    }
  }
}

export const deleteNote = (id) => ({
  type: types.notesDelete,
  payload: id
})

export const notesLogout = () => ({
  type: types.notesLogoutCleaning
})

