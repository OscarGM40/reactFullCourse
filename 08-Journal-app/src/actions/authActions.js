import { firebase, googleAuthProvider } from "../firebase/firebase-config";
import { types } from "../types/types";
import { startLoading, finishLoading } from "../actions/uiActions";
import Swal from "sweetalert2";
import { notesLogout } from "./notesActions";

// fijate que algunos si tienen la lógica(los asincronos) mientras que los sincronos( no sé si son los de redux también) solo declaran la accion.Es un poco confusa esta forma.
export const startLoginEmailPassword = (email, password) => {
  return async (dispatch) => {
    dispatch(startLoading());

    try {
      const { user } = await firebase
        .auth()
        .signInWithEmailAndPassword(email, password);
      dispatch(login(user.uid, user.displayName));
      dispatch(finishLoading());
    } catch (e) {
      console.log("error en el login");
      dispatch(finishLoading());
      Swal.fire("Error", e.message, "error");
      return e.code;
    }
  };
};

export const startRegisterWithEmailPasswordName = (email, password, name) => {
  return async (dispatch) => {
    // firebase trabajará con promesas,ojo
    try {
      const userCred = await firebase
        .auth()
        .createUserWithEmailAndPassword(email, password);
      const { user } = userCred;
      // es muy conveniente acceder a este método y subir el nombre también
      await user.updateProfile({ displayName: name });
      // console.log(user)
      dispatch(login(user.uid, user.displayName));
    } catch (e) {
      console.log("error en el registro");
      Swal.fire("Error", e.message, "error");
      return e.code;
    }
  };
};

export const startGoogleLogin = () => {
  // con esta forma thunk provee del dispatch en el return,es por esto que puedo usarlo
  //recuerda que esta forma con el return es por usar redux y no @reduxjs/toolkit.No es díficil,sólo cambia un poco
  return (dispatch) => {
    //singInWithPopup pide un proveedor y es una promesa
    firebase
      .auth()
      .signInWithPopup(googleAuthProvider)
      // la función resuelve a un UserCredential
      .then((userCred) => {
        const { user } = userCred;
        dispatch(login(user.uid, user.displayName));
      });
  };
};

export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    displayName,
  },
});

// si bien es un poquito más liosa esta forma también parece interesante.
export const startLogout = () => {
  // es asincrono porque hay comunicacion con un backend asin
  // fijate que puedo usar async en el return y trabajar con await en el interior.Realmente return sólo indica que es una instrucción terminal.
  return async (dispatch) => {
    // si bien es una Promise no se suele usar el catch pues no suele fallar
    await firebase.auth().signOut();
    dispatch(logout());
    dispatch(notesLogout());
  };
};

export const logout = () => ({
  type: types.logout,
});
