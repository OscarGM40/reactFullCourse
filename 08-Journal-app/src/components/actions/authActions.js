import { firebase, googleAuthProvider } from "../firebase/firebase-config";
import { types } from "../../types/types";
import { startLoading, finishLoading } from "../actions/uiActions";
import Swal from "sweetalert2";

// fijate que algunos si tienen la lógica(los asincronos) mientras que los sincronos( no sé si son los de redux también) solo declaran la accion.Es un poco confusa esta forma.
export const startLoginEmailPassword = (email, password) => {
  return (dispatch) => {
    dispatch(startLoading());
    firebase
      .auth()
      .signInWithEmailAndPassword(email, password)
      .then(({ user }) => {
        dispatch(login(user.id, user.displayName));
        dispatch(finishLoading());
      })
      .catch((e) => {
        console.log('por aqui');
        dispatch(finishLoading());
        Swal.fire("Error", e.message, "error");
      });
  };
};

export const startRegisterWithEmailPasswordName = (email, password, name) => {
  return (dispatch) => {
    // firebase trabajará con promesas,ojo
    firebase
      .auth()
      .createUserWithEmailAndPassword(email, password)
      .then(async (userCred) => {
        const { user } = userCred;
        // es muy conveniente acceder a este método y subir el nombre también
        await user.updateProfile({ displayName: name });
        // console.log(user)
        dispatch(login(user.id, user.displayName));
      })
      .catch((e) => {
        console.log(e);
        Swal.fire("Error", e.message, "error");
      });
    };
  };
  
  export const startGoogleLogin = () => {
    // con esta forma thunk provee del dispatch en el return,es por esto que puedo usarlo
    //recuerda que esta forma con el return es por usar redux y no @reduxjs/toolkit.No es díficil,sólo cambia un poco
    return (dispatch) => {
      //singInWithPopup pide un proveedor y es una promesa
      firebase
      .auth()
      .signInWithPopup(googleAuthProvider)
      // la función resuelve a un UserCredential
      .then((userCred) => {
        const { user } = userCred;
        dispatch(login(user.uid, user.displayName));
      });
    };
  };
  
  export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
      uid,
      displayName,
    },
  });
  
  // si bien es un poquito más liosa esta forma también parece interesante.
  export const startLogout = () => {
    // es asincrono porque hay comunicacion con un backend asin
    // fijate que puedo usar async en el return y trabajar con await en el interior.Realmente return sólo indica que es una instrucción terminal.
    return async (dispatch) => {
      // si bien es una Promise no se suele usar el catch pues no suele fallar
      await firebase.auth().signOut();
      dispatch(logout);
    };
  };
  
  export const logout = () => ({
    type: types.logout,
  });
  