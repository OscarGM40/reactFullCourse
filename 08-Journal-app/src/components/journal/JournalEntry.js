import React from "react";
import moment from "moment";
// si quiero la traduccion importar esto ↓↓
// import "moment/locale/es";
import { useDispatch } from "react-redux";
import { activateNote } from "../actions/notesActions"



export const JournalEntry = ({ id, title, body, date,url }) => {

  const dispatch = useDispatch()
  const noteDate = moment(date);

  const setActiveNote = () => {
      dispatch(activateNote(id,{title,body,date,url}));
  }
  
  
  return (
    <div className="journal__entry pointer animate__animated animate__flipInY animate__fast"
      onClick={setActiveNote}>
      {
        url &&
        <div
          className="journal__entry-picture"
          style={{
            backgroundSize: "cover",
            backgroundImage: `url(${url})`,
          }}
        >
        </div>
      }
      <div className="journal__entry-body">
        <p className="journal__entry-title">{title}</p>
        <p className="journal__entry-content">{body}</p>
      </div>

      <div className="journal__entry-date-box">
        <span>{ noteDate.format("dddd") }</span>
        <h4>{ noteDate.format("Do") }</h4>
      </div>
    </div>
  );
};
