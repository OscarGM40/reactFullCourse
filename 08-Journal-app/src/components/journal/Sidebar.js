import React from "react";
import { useSelector } from "react-redux";
import { useDispatch } from "react-redux";
import { startLogout } from "../../actions/authActions";
import { startNewNote } from "../../actions/notesActions";
import { JournalEntries } from "./JournalEntries";

export const Sidebar = () => {
  

  const dispatch = useDispatch()
  const { name } = useSelector( state => state.auth) 

  // fijate que con esta forma disparo aqui la accion y en el otro file este método a su vez también dispara tras la asincronia el borrado de redux.Interesante,realmente puede programarse lo que se quiera
  const handleLogout = () => {
    dispatch(startLogout());
    // setTimeout(() => window.location.href="/auth/login",300)
  }

  const handleAddNew = () => {
    dispatch(startNewNote());
  }
  
  return (
    <aside className="journal__sidebar">
      <div className="journal__sidebar-navbar">
        <h3 className="mt-5">
          <i className="far fa-moon"></i>
          <span> {name}</span>
        </h3>

        <button className="btn"
          onClick={handleLogout}>Logout</button>
      </div>

      <div className="journal__new-entry"
        onClick={handleAddNew}>
        <i className="far fa-calendar-plus fa-5x "></i>
        <p className="mt-5">New Entry</p>
      </div>

      <JournalEntries />
      
    </aside>
  );
};
