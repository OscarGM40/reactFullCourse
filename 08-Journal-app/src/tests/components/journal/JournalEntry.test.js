import "@testing-library/jest-dom";
import React from "react";
import { mount } from "enzyme";
import { Provider } from "react-redux";

import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

import { JournalEntry } from "../../../components/journal/JournalEntry";
import { types } from "../../../types/types";
import { activateNote } from "../../../actions/notesActions";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initialState = {};

let store = mockStore(initialState);
store.dispatch = jest.fn(); //usar esta funcion me permite ver con qué se disparó el dispatch

// recuerda que si un componente recibe props se las tengo que pasar simulandolas
const nota = {
  id: 10,
  date: 0,
  title: "hola",
  body: "mundo",
  url: "https://algunlugarl.com/foto.jpg",
};
const wrapper = mount(
  <Provider store={store}>
    <JournalEntry {...nota} />
  </Provider>
);

describe("Pruebas en JournalEntry", () => {

  /*   beforeEach(() => {
    store = mockStore(initialState);
    store.clearActions();
  }); */

  test("debe de mostrarse correctamente", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("debe de activar la nota correctamente", async () => {

    wrapper.find("div.journal__entry").prop("onClick")();
    
    expect(store.dispatch).toHaveBeenCalled();
    expect(store.dispatch).toHaveBeenCalledWith({
      type: types.notesActive,
      payload: {
        id: 10,
        title: "hola",
        body: "mundo",
        date: 0,
        url: "https://algunlugarl.com/foto.jpg",
      },
    }); 
    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(
      activateNote(nota.id,{...nota}));
  });

});
