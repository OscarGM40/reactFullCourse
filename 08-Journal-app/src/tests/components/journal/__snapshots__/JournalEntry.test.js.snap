// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en JournalEntry debe de mostrarse correctamente 1`] = `
<div
  className="journal__entry pointer animate__animated animate__flipInY animate__fast"
  onClick={[Function]}
>
  <div
    className="journal__entry-picture"
    style={
      Object {
        "backgroundImage": "url(https://algunlugarl.com/foto.jpg)",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      hola
    </p>
    <p
      className="journal__entry-content"
    >
      mundo
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
      Thursday
    </span>
    <h4>
      1st
    </h4>
  </div>
</div>
`;
